packages/labs/src/components/select/multiSelect.tsx

```diff
@@ -0,0 +1,256 @@
+
+    /**
+     * Whether the popover opens on key down or when `TagInput` is focused.
+     * @default false
+     */
+    openOnKeyDown?: boolean;
+
+    /** Props to spread to `Popover`. Note that `content` cannot be changed. */
+    popoverProps?: Partial<IPopoverProps>;
+
+    /**
+     * Whether the filtering state should be reset to initial when an item is selected
+     * (immediately before `onItemSelect` is invoked), or when the popover closes.
+     * The query will become the empty string and the first item will be made active.
+     * @default true
+     */
+    resetOnSelect?: boolean;
+
+    /** Props to spread to `TagInput`. */
+    tagInputProps?: Partial<ITagInputProps>;
+
+    /** Custom renderer to transform an item into a string for tags. */
+    tagRenderer: (item: T) => string;
+}
+
+export interface IMultiSelectState<T> {
+    activeItem?: T;
+    isOpen?: boolean;
+    query?: string;
+}
+
+@PureRender
+export class MultiSelect<T> extends AbstractComponent<IMultiSelectProps<T>, IMultiSelectState<T>> {
+    public static displayName = "Blueprint.MultiSelect";
+
+    public static ofType<T>() {
+        return MultiSelect as new () => MultiSelect<T>;
+    }
+
+    public state: IMultiSelectState<T> = {
+        isOpen: false,
+        query: "",
+    };
+
+    private TypedQueryList = QueryList.ofType<T>();
+    private input: HTMLInputElement;
+    private queryList: QueryList<T>;
+    private queryListKeyDown: React.EventHandler<React.KeyboardEvent<HTMLElement>>;
+    private refHandlers = {
+        input: (ref: HTMLInputElement) => this.input = ref,
+        queryList: (ref: QueryList<T>) => this.queryList = ref,
+    };
+
+    public render() {
+        // omit props specific to this component, spread the rest.
+        const {
+            itemRenderer,
+            noResults,
+            openOnKeyDown,
+            popoverProps,
+            resetOnSelect,
+            tagInputProps,
+            ...props,
+        } = this.props;
+
+        return <this.TypedQueryList
+            {...props}
+            activeItem={this.state.activeItem}
+            onActiveItemChange={this.handleActiveItemChange}
+            onItemSelect={this.handleItemSelect}
+            query={this.state.query}
+            ref={this.refHandlers.queryList}
+            renderer={this.renderQueryList}
+        />;
+    }
+
+    private renderQueryList = (listProps: IQueryListRendererProps<T>) => {
+        const { tagInputProps = {}, popoverProps = {} } = this.props;
+        const { handleKeyDown, handleKeyUp, query } = listProps;
+        const defaultInputProps: HTMLInputProps = {
+            placeholder: "Search...",
+            ...this.props.tagInputProps.inputProps,
+            // tslint:disable-next-line:object-literal-sort-keys
+            onChange: this.handleQueryChange,
+            ref: this.refHandlers.input,
+            value: query,
+        };
+
+        this.queryListKeyDown = handleKeyDown;
```

Comment:
meh yeah this is kind of hairy. would prefer to make an exception to the jsx lambda rule here and refactor `this.handleKeyDown` into a getter:

```ts
    <div
        onKeyDown={this.getTargetKeyDownHandler(listProps.handleKeyDown)}
        onKeyUp={this.state.isOpen && handleKeyUp}
    >

...

    private getTargetKeyDownHandler(handleQueryListKeyDown: React.EventHandler<React.KeyboardEvent<HTMLElement>) {
        return (e: React.KeyboardEvent<HTMLElement>) => {
            const { which } = e;
            const { resetOnSelect } = this.props;

            ...

            if (this.state.isOpen) {
                Utils.safeInvoke(handleQueryListKeyDown, e);
            }
        };
    }
```